synapse/util/caches/descriptors.py

```diff
@@ -208,26 +207,12 @@ def _wrapped(*args, **kwargs):
                 kwargs["cache_context"] = _CacheContext.get_instance(cache, cache_key)
 
             try:
-                cached_result_d = cache.get(cache_key, callback=invalidate_callback)
-
-                if isinstance(cached_result_d, ObservableDeferred):
-                    observer = cached_result_d.observe()
-                else:
-                    observer = defer.succeed(cached_result_d)
-
+                ret = cache.get(cache_key, callback=invalidate_callback)
             except KeyError:
                 ret = defer.maybeDeferred(preserve_fn(self.orig), obj, *args, **kwargs)
+                cache.set(cache_key, ret, callback=invalidate_callback)
 
-                def onErr(f):
-                    cache.invalidate(cache_key)
-                    return f
-
-                ret.addErrback(onErr)
```

Comment:
this happens anyway, at https://github.com/matrix-org/synapse/blob/e67f5720944687ebed524408c08076b35178f2b3/synapse/util/caches/deferred_cache.py#L264.

Reply:
ohh, except that's slightly different. I better write some tests.

Reply:
ok, I've stared at this harder and am pretty sure it's redundant. As noted above, if the function fails, then we call `entry.invalidate()` on the "temporary" entry in the pending_deferred_cache. The only difference between that and calling `cache.invalidate(cache_key)` as shown here would be if there was also an entry in the *main* cache. But we know that can't be the case, because we checked it at line 210, and there is no way for it to have been populated in the meantime.

https://github.com/matrix-org/synapse/pull/8572/files#diff-3aa4bc9d8830ffd035d53b8af86b548a21f2a4cddb3e5240445661614a88a08fR210

(unless someone decided to sidestep the CacheDescriptor and manually call `cache.set` or `cache.prefill` while this request was running? but that would be silly, and even if they did it's not obvious that we should invalidate it here.)

In short, I think this is handled correctly in DeferredCache itself.